comp/core/workloadmeta/collectors/internal/ecs/ecs.go

```diff
@@ -233,6 +276,126 @@ func (c *collector) parseTaskContainers(
 	return taskContainers, events
 }
 
+func (c *collector) parseV4Task(ctx context.Context, task v1.Task, seen map[workloadmeta.EntityID]struct{}) []workloadmeta.CollectorEvent {
+	v4Task := c.getV4TaskWithTags(ctx, task)
+	return util.ParseV4Task(v4Task, seen)
+}
+
+// getV4TaskWithTags fetches task and tasks from the metadata v4 API
+func (c *collector) getV4TaskWithTags(ctx context.Context, task v1.Task) v3or4.Task {
+	var metaURI string
+	for _, taskContainer := range task.Containers {
+		containerID := taskContainer.DockerID
+		container, err := c.store.GetContainer(containerID)
+		if err != nil {
+			log.Tracef("cannot find container %q found in task %s: %s", taskContainer, task.Arn, err)
+			continue
+		}
+
+		uri, ok := container.EnvVars[v3or4.DefaultMetadataURIv4EnvVariable]
+		if ok && uri != "" {
+			metaURI = uri
+			break
+		}
+	}
+
+	if metaURI == "" {
+		log.Errorf("failed to get client for metadata v3 or v4 API from task %s and the following containers: %v", task.Arn, task.Containers)
```

Comment:
If, in this and any of the other error conditions, we return the task, wouldn't we would try again in the future? Could we end up in a state where a task always fails, and we keep retrying?

Reply:
Yes, we will retry it in the next pull (interval by default 5s)
https://github.com/DataDog/datadog-agent/blob/882d6e4354906f23c0124aa0d908eed0e5bcbd4a/comp/core/workloadmeta/collectors/internal/ecs/ecs.go#L125

> Could we end up in a state where a task always fails, and we keep retrying?

It shouldn't happen as long as docker collector works fine